{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BACKGROUND INFORMATION\n",
    "\n",
    "I have retrieved the Locus Germplasm Phenotype file from The Arabidiosis Information Resource (TAIR - one of the primary _Arabidopsis_ databases globally) (https://www.arabidopsis.org/download/index-auto.jsp?dir=/download_files/Polymorphisms_and_Phenotypes ).  The file contains information about the phenotypes associated with certain _Arabodopsis_ mutations.\n",
    "\n",
    "The file is in the current folder,named: [Locus_Germplasm_Phenotype_20130122.txt](./Locus_Germplasm_Phenotype_20130122.txt)\n",
    "\n",
    "The file format is \"tab-separated values\".  This means that each data field is separated by a \"tab\" character.  \n",
    "\n",
    "**There is one line at the beginning of the file - the \"header\" - that describes the type of data in each column.**  \n",
    "\n",
    "The first column is the \"Arabiodopsis Genome Initiative\" (AGI) Locus Code.  Locus Codes have a predictable structure\n",
    "\n",
    "* The first character is \"A\" or \"a\"\n",
    "* The second character is \"T\" or \"t\"\n",
    "* The third character is the chromosome number (between 1 and 5)\n",
    "* The fourth character is \"G\" or \"g\"\n",
    "* The remaining characters are a set of 5 digits (between 0 and 9)\n",
    "\n",
    "**For Example:  At5G22934**\n",
    "\n",
    "The final column in the dataset is the PubMed ID for the published paper describing this phenotype.  PubMed IDs are all digits (0-9) but the length of the ID is unpredictable.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Exam Problems - Answers must be submitted by 1800h Monday, Sept 6.\n",
    "\n",
    "**You may use ANY RESOURCE YOU WANT - including Google, or the course notes.  **\n",
    "\n",
    "**Please DO NOT discuss your answers with each other until after the exam is complete.**\n",
    "\n",
    "You will submit your answers as a Jupyter Notebook.\n",
    "\n",
    "For each problem #2 to #10 create a \"Markdown box\" explaining your answer, and a \"code box\" containing the command(s)  (problems #1 and #11 do not require explanation)\n",
    "\n",
    "Last week, you \"forked\" the Wilkinson Exams git (Accelerated_Intro_WilkinsonExams) into your own github account.  You will use your Git to store the answers to the exams.\n",
    "\n",
    "\n",
    "## Problem 1\n",
    "\n",
    "\n",
    "* Copy the Locus_Germplasm_Phenotype_20130122.txt file into your copy of the Exams Git\n",
    "* Create a new Jupyter Notebook in your Exams folder called \"Exam_1_Answers\" using the 'bash' kernel \n",
    "* Do the rest of this exam inside of that the \"Exam_1_Answers\" Notebook\n",
    "\n",
    "\n",
    "## Problem 2\n",
    "\n",
    "Create a directory listing command that shows:\n",
    "\n",
    "* the ownership of the file\n",
    "* it's file-size, in Megabytes (i.e. human readable)\n",
    "\n",
    "then:\n",
    "\n",
    "* in words (in a Markdown box), describe the _permissions_ on that file (read/write/execute) for users, groups, and \"anyone\"\n",
    "\n",
    "\n",
    "## Problem 3\n",
    "\n",
    "* Create a command that outputs only the \"header\" line of Locus_Germplasm_Phenotype_20130122.txt\n",
    "\n",
    "\n",
    "## Problem 4\n",
    "\n",
    "* Create a command that outputs the total number of lines in Locus_Germplasm_Phenotype_20130122.txt\n",
    "\n",
    "\n",
    "## Problem 5\n",
    "\n",
    "* Create a command that writes ONLY the data lines (i.e. excludes the header!) to a new file called \"Data_Only.csv\"\n",
    "* prove that your output file has the expected number of lines\n",
    "\n",
    "\n",
    "## Problem 6\n",
    "\n",
    "* Create a command that shows all of the lines that have a phenotype including the word \"root\"\n",
    "\n",
    "## Problem 7\n",
    "\n",
    "* Create a command that writes the AGI Locus Code for every line that has a phenotype including the word \"root\" to a file called: **Root-associated-Loci.txt**\n",
    "\n",
    "\n",
    "## Problem 8\n",
    "\n",
    "* Create a command that writes the PubMed ID for every line that has a phenotype including the word \"root\" to a file called: **Root-associated-Publications.txt**\n",
    "\n",
    "\n",
    "## Problem 9\n",
    "\n",
    "* _**Control experiment**_:  You would hypothesize that genes associated with roots **should be found on all chromosomes.**  Find a way (one or more commands) to test this hypothesis.  In this dataset, is the hypothesis true? \n",
    "\n",
    "\n",
    "## Problem 10\n",
    "\n",
    "* If your control experiment shows genes on every chromosome, then you can skip this question!  (you answered Problem 9 correctly!)\n",
    "\n",
    "*  If your control experiment shows genes only on one or two chromosomes, then you have to explain why... what could the problem be?  (I told you specifically to be careful about this problem!)\n",
    "\n",
    "\n",
    "## Problem 11\n",
    "\n",
    "* 'git commit' and 'git push' your answers to your GitHub, then give me your GitHub username before you leave the class. I will clone your repositories and grade your answers.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
