{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "## Start your mySQL Server from a terminal \n",
    "(if it isn't already running)\n",
    "\n",
    "<code>sudo docker start course-mysql</code>\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "**Don't forget that, if you use sqlMagic, you need to execute the connection lines in your Notebook!**\n",
    "\n",
    "<pre>\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "</pre>\n",
    "\n",
    "## Create a new Python3 Jupyter Notebook in your Exam Answers folder\n",
    "\n",
    "Commit and push this Notebook to GitHub when you are finished.\n",
    "\n",
    "You must submit your answers to GitHub by 1800h Sept 14."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Data Files\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n"
     ]
    }
   ],
   "source": [
    "import csv \n",
    "\n",
    "germplasmfile = open(\"Germplasm.tsv\", \"r\")\n",
    "locusgenefile = open(\"LocusGene.tsv\", \"r\")\n",
    "\n",
    "germplasmfile.seek(0)\n",
    "locusgenefile.seek(0)\n",
    "\n",
    "listalocus=[]\n",
    "i=0\n",
    "\n",
    "for line1 in germplasmfile.readlines():\n",
    "    listalocus.append(line1.split(\"\\t\")[0])\n",
    "\n",
    "for line2 in locusgenefile.readlines():\n",
    "    line2 = line2.split(\"\\t\")[0]\n",
    "    \n",
    "    if listalocus[i] == line2:\n",
    "        print(\"match\")\n",
    "    else:\n",
    "        print(\"no match\")\n",
    "    \n",
    "    i=i+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a **1:1 relationship**\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "            \n",
    "%sql CREATE DATABASE Exam2_database;           \n",
    "%sql use Exam2_database;\n",
    "\n",
    "%sql CREATE TABLE germplasm_table(Locus VARCHAR(9) NOT NULL PRIMARY KEY, Germplasm VARCHAR(50) NOT NULL PRIMARY KEY, Phenotype VARCHAR(1500) NOT NULL, Pubmed VARCHAR(50) NOT NULL); \n",
    "%sql CREATE TABLE locus_gene_table(Locus VARCHAR(9) NOT NULL PRIMARY KEY, Gene VARCHAR(20) NOT NULL, Protein_Length VARCHAR (20) NOT NULL)\n",
    "\n",
    "#%sql DROP TABLE germplasm_table;\n",
    "#%sql DROP TABLE locus_gene_table;\n",
    "#%sql show tables\n",
    "#%sql describe germplasm_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1146, \"Table 'Exam2_database.germplasm_table' doesn't exist\")\n",
      "[SQL: describe germplasm_table]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n"
     ]
    }
   ],
   "source": [
    "%sql describe germplasm_table\n",
    "#%sql describe locus_gene_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and **fills the database**.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'Locus_Gene.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-f01122a4379f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgermplasm_table\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Germplasm.tsv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mlocus_gene_table\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Locus_Gene.tsv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mlocus_genelist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mgermplasmlist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'Locus_Gene.tsv'"
     ]
    }
   ],
   "source": [
    "germplasm_table = open(\"Germplasm.tsv\", \"r\")\n",
    "locus_gene_table = open(\"Locus_Gene.tsv\", \"r\")\n",
    "locus_genelist = []\n",
    "germplasmlist = []\n",
    "\n",
    "for line in locus_gene.readlines():\n",
    "    line = line.rstrip()\n",
    "    Locus, Gene, ProteinLength = line.split('\\t')\n",
    "    locus_genelist.append((Locus, Gene, ProteinLength)) #I do the list inserting data with .append() command\n",
    "for line in germplasm.readlines():\n",
    "    line = line.rstrip()\n",
    "    Locus, Germplasm, Phenotype, Pubmed = line.split('\\t')\n",
    "    germplasmlist.append((Locus, Germplasm, Phenotype, Pubmed))\n",
    "locusgene.close()\n",
    "germplasm.close()\n",
    "\n",
    "locusgenelist.pop(0)\n",
    "germplasmlist.pop(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, **content of the two database tables (including a header line)**\n",
    "\n",
    "2. Create a joined report that only includes the Genes **SKOR and MAA3**\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome **(AT1Gxxxxxx to AT5Gxxxxxxx)**\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome **(AT1Gxxxxxx to AT5Gxxxxxxx)**\n",
    "\n",
    "**When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule!** \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "</pre>\n",
    "## Don't forget to commit and push your answers before you leave!\n",
    "\n",
    "It was wonderful to have you in my class!  I hope to see you again soon!\n",
    "\n",
    "Good luck with your careers!!\n",
    "\n",
    "Mark\n",
    "\n",
    "\n",
    "**THANK YOU VERY MUCH FOR YOUR CLASSES MARK! See you in the course Challenges in Computer Programming!!**\n",
    "\n",
    "**Marina**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
